resynctool.8
.\" Automatically generated by Pod::Man 2.16 (Pod::Simple 3.07)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "resynctool 8"
.TH resynctool 8 "2008-02-11" "otpd 3.2.4" " "
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
resynctool \- Manual resync tool for event\-based HOTP tokens
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
resynctool \-1 \fIotp1\fR \-2 \fIotp2\fR [options]
.IP "\fB\-1\fR \fIotp1\fR		first \s-1OTP\s0 \fIotp1\fR" 4
.IX Item "-1 otp1		first OTP otp1"
.PD 0
.IP "\fB\-2\fR \fIotp2\fR		second \s-1OTP\s0 \fIotp2\fR" 4
.IX Item "-2 otp2		second OTP otp2"
.IP "\fB\-u\fR \fIuser\fR		emit state for user \fIuser\fR" 4
.IX Item "-u user		emit state for user user"
.IP "\fB\-k\fR \fIkey\fR		use key \fIkey\fR" 4
.IX Item "-k key		use key key"
.IP "\fB\-c\fR \fIchallenge\fR	set initial counter value from \fIchallenge\fR" 4
.IX Item "-c challenge	set initial counter value from challenge"
.IP "\fB\-i\fR \fIcounter\fR	initial \fIcounter\fR value to test" 4
.IX Item "-i counter	initial counter value to test"
.IP "\fB\-f\fR \fIcounter\fR	final \fIcounter\fR value to test" 4
.IX Item "-f counter	final counter value to test"
.IP "\fB\-d\fR			debug switch" 4
.IX Item "-d			debug switch"
.PD
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
Event-based tokens generate an \s-1OTP\s0 by using an \fBevent counter\fR as input
to the \s-1OTP\s0 algorithm.  The event counter is incremented on each use of
the token.
.PP
Unlike time-based tokens, event-based tokens can get out of sync with the
server due to network issues or misuse.  The \fIrwindow_size\fR setting in
\&\fI/etc/otpd.conf\fR is used to search over a larger range of event counter
values (starting from the last known good counter) than normal, and can
be used to overcome sync problems to some extent, but it is not advised
to set it very large as it creates more work for the server, and therefore
with \*(L"severe\*(R" sync problems the user may become effectively locked out.
.PP
When using \s-1LDAP\s0 with \fBotpd\fR, a user-specific \fIrwindow_size\fR setting
can be made, which allows the user to self-resync.  The TRI-D Systems
Professional Edition software also comes with a web application that
administrators or users can use to make this job easy.
.PP
When using files with \fBotpd\fR, user-specific settings are not available,
and therefore \fBresynctool\fR is provided.  Some sites may want to use it
with \s-1LDAP\s0 as well.
.PP
To use \fBresynctool\fR, give the username, token key and two consecutive
freshly generated OTPs.  \fBresynctool\fR will output the state data which
will resync the user.  Suggested usage (wrap this in a script):
.PP
.Vb 2
\&  resynctool \-u USER \-k ... \-1 123456 \-2 123456 >/dev/null &&
\&  resynctool \-u USER \-k ... \-1 123456 \-2 123456 >/etc/otpstate/USER
.Ve
.PP
\&\fBresynctool\fR is designed to work with TRI-D event-based \s-1HOTP\s0 tokens,
however it will work with any event-based \s-1HOTP\s0 token which uses a counter
that increments by one for each subsequent use of the token.  (\s-1RFC\s0 4226
does not specify how the counter should be derived.)
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\fB\-1\fR \fIotp1\fR" 4
.IX Item "-1 otp1"
The first of two consecutive OTPs.
.IP "\fB\-2\fR \fIotp2\fR" 4
.IX Item "-2 otp2"
The second of two consecutive OTPs.
.IP "\fB\-u\fR \fIuser\fR" 4
.IX Item "-u user"
The user that holds the token; it is necessary for the state data.
If not provided, a generic value will be used which must then be
replaced with the actual username, otherwise \fBotpd\fR will not
consider the state to be valid.
.IP "\fB\-k\fR \fIkey\fR" 4
.IX Item "-k key"
The token secret.
.IP "\fB\-c\fR \fIchallenge\fR" 4
.IX Item "-c challenge"
The last known good challenge (hex counter) value.  It is an 8
octet value (16 hex digits), stored as the third field of v5
state.  Using this option tells \fBresynctool\fR to begin searching
for candidate event counters at this value instead of at 0.  This
saves work testing event counters that cannot possibly be valid.
To extract the challenge value (work this into your wrapper script):
.Sp
.Vb 1
\&  cut \-d: \-f3 /etc/otpstate/USER
.Ve
.IP "\fB\-i\fR \fIcounter\fR" 4
.IX Item "-i counter"
This is an alternative to the \fB\-c\fR option.  Instead of a hex
value, this option takes a decimal counter.
.IP "\fB\-f\fR \fIcounter\fR" 4
.IX Item "-f counter"
The final (decimal) counter value to test.  By default, \fBresynctool\fR
will test 65536 values before giving up.  If the first character of
\&\fIcounter\fR is \f(CW\*(C`+\*(C'\fR (plus), instead of using \fIcounter\fR as an absolute
value it will be added to the initial value.
.IP "\fB\-d\fR" 4
.IX Item "-d"
Activate debug output (ex: the counters)
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIotpd\fR\|(8), \s-1RFC\s0 4226
.SH "URL"
.IX Header "URL"
<http://otpd.googlecode.com/>


NEWS
otpd-3.1.0: 2008/02/11
  - gsmd: add support for gsmd backup server and failover
  - gsmd: fix request type field in gsmd message
  - gsmd: combine GET and PUT seqno sequences into a single sequence
  - lsmd: guarantee that entire state data is read
  - lsmd: report error if state data is too large
  - lsmd: reduce max state size to 892 bytes from 1024 bytes
  - lsmd: improve reporting of statedir errors
  - config: disallow gsmd configuration in otpd.conf

otpd-3.0.0: 2008/01/30
  - integrate lsmd (lsmd is now eliminated)
  - fix timeout handling affecting gsmd and LDAP communication (timer ran fast)
  - update userops API
  - add gsmd support when using file backend
  - config: fix test for file vs ldap mutual exclusivity
  - config: add explicit backend keyword

otpd-2.5.2: 2007/12/12
  - fix per-user rwindow bug (random value) when using file backend

otpd-2.5.1: 2007/12/11
  - fix ascii to hex conversion bug introduced in 2.5.0
  - fix otpauth error string reporting

otpd-2.5.0: 2007/12/07
  - add resynctool for manual resync of event-based hotp tokens
  - change ewindow and rwindow defaults to 5 and 25, respectively
  - remove cyclades support
  - ldap: add per-user tmp static passwd override
  - ldap: add per-user tmp rwindow override
  - ldap: Active Directory schema fixes
  - userops: fix ctor declaration on Solaris
  - configure: check for $CC tls support
  - build: link otpauth against minimal set of libraries
  - install: save old config file

otpd-2.4.0: 2007/11/09
  - rate limit rwindow auths to 1/s
  - update documentation
  - update packaging
  - recommend 25 for rwindow
  - ldap: add Active Directory support
  - trid: fix little-endian event-based HOTP bug

otpd-2.3.0: 2007/10/02
  - add OTP_RC_NEXTPASSCODE to inform client of rwindow candidate action
  - add pin-md5 encrypt mode
  - trid: remove beta-3 (final beta) support
  - trid: support event-based HOTP tokens
  - update license (trivial typo)

otpd-2.2.1: 2007/07/12
  - trid: fix nullstate bug (nullstate didn't work correctly)

otpd-2.2.0: 2007/05/22
  - fix off-by-one bug in key configuration, which disallowed sequential keyids
  - change generic hotp card names (add a '-') to support LDAP
  - change otp_request_t to v2, to save some space
  - cardops: update name2fm() method to accept state arg
  - ldap: update otpSerialPool from STRUCTURAL to AUXILIARY
  - trid: remove suppport for 5-digit OTPs
  - trid: improve card name parsing
  - trid: add totp (TRI-D OTP) algorithm
  - trid: add new card naming format, to select hard/soft and hotp/totp
  - trid: deprecate trid-hotp-* names in favor of trid-hard-hotp-*
  - trid: remove beta-2 support

otpd-2.1.8: 2007/05/03
  - ldap: more robust handling of otpKeyEncryption errors
  - ldap: allow otpKeyEncryption=clear, as documented
  - ldap: update otp.schema for openldap < 2.3 compatibility
  - ldap: correct otp.schema closing parens syntax
  - include otpauth utility
  - fix Makefile install target

otpd-2.1.7: 2007/04/18
  - ldap: link statically against included libldap
  - ldap: add otpTokenPool objectclass

otpd-2.1.6: 2007/04/16
  - ldap: fix included libldap missing symlink

otpd-2.1.5: 2007/04/15
  - fix build of userops modules
  - ldap: include libldap

otpd-2.1.4: not released

otpd-2.1.3: 2007/04/13
  - fix cryptocard and x99 "phone response" passcode length bug
  - actually accept -u option as documented
  - ldap: fix some minor config error reporting bugs
  - ldap: change otpToken objectClass from STRUCTURAL to AUXILIARY
  - ldap: release ldap module as open source

otpd-2.1.2: 2007/01/19
  - fix state release (PUT) command formatting in error path (rare issue)
  - ldap: add otpDuressAlert object, rename otpAuthAlertAddress to otpAuthAlert

otpd-2.1.1: 2007/01/13
  - set pin correctly in pin encrypt mode
  - fix buffer overflow in key decrypt when using pin encrypt mode

otpd-2.1: 2007/01/12
  - add otppasswd key and pin encryption
  - change otppasswd pin format from ASCII to hex
  - early exit (skip crypto) for plaintext auth if passcode length is wrong
  - don't run cardops and userops constructors more than once on solaris
  - fix lsmd/gsmd config ignore; handle any level of subsections
  - improve config syntax checking of otpd section
  - cardops: rename keystring2keyblock method as keystring2key
  - cardops: simplify name2fm method
  - userops: return error if PIN format is wrong
  - ldap: handle missing otpPIN and otpStateLocationHint correctly
  - ldap: fix memleak in attribute format error path
  - update API docs

otpd-2.0: 2007/01/04
  - ldap: add TLS support
  - ldap: add timeout when connecting to ldap server
  - ldap: use connection pool to allow re-use
  - ldap: require OpenLDAP >= 2.3.4, for start tls (drops FC3 and FC4 support)
  - ldap: validate otpTimeIssued attribute
  - ldap: don't try to parse error result
  - ldap: fix bug allowing only ridiculously short hostnames
  - ldap: add otpAuthAlertAddress object
  - userops: add put() method
  - userops: allocate user in get() (rather than having caller allocate)
  - fix pointer target signedness errors for gcc-4.x and Sun CC

otpd-1.99.2: 2006/12/20 (limited test release)
  - add timeout config var, and propagate it (smartly) to lsmd
  - remove ldap timeout config var (use top level timeout)
  - send gsmd data to lsmd
  - fix state manager request buf undersize bug
  - fix ascii vs hex bug when saving cryptocard challenge data
  - fix lsmd/gsmd config parsing to ignore subsections
  - syslog reason for GET NAK response from state manager
  - syslog (at LOG_NOTICE) when user auth fails
  - change user auth success syslog to LOG_NOTICE from LOG_INFO
  - improve logging for user_get() failures
  - log starting/stopping message with progname and version
  - update API documentation
  - userops: run all known constructors if they didn't run (for Cyclades ACS)
  - userops: add hrtime arg to get() method
  - ldap: return error on lookup errors, don't exit

resynctool.c
/*
 * resynctool.c
 * find the counter value for trid event-based hotp tokens
 * prints a state file entry to stdout
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 *
 *  For alternative licensing terms, contact licensing@tri-dsystems.com.
 *
 * Copyright 2005-2007 TRI-D Systems, Inc.
 */

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <openssl/hmac.h>
#include <unistd.h>

extern char *optarg;

static void
usage(const char *progname)
{
  fprintf(stderr, "usage: %s -1 <otp1> -2 <otp2> [-u <user>] [-k <key>] [-c <challenge>] [-i <counter>] [-f <counter>] [-d]\n", progname);
  exit(1);
}

static ssize_t
a2x(const char *s, unsigned char x[])
{
  unsigned i;
  size_t l = strlen(s);

  /*
   * We could just use sscanf, but we do this a lot, and have very
   * specific needs, and it's easy to implement, so let's go for it!
   */
  for (i = 0; i < l / 2; ++i) {
    unsigned int n[2];
    int j;

    /* extract 2 nibbles */
    n[0] = *s++;
    n[1] = *s++;

    /* verify range */
    for (j = 0; j < 2; ++j) {
      if ((n[j] >= '0' && n[j] <= '9') ||
          (n[j] >= 'A' && n[j] <= 'F') ||
          (n[j] >= 'a' && n[j] <= 'f'))
        continue;
      return -1;
    }

    /* convert ASCII hex digits to numeric values */
    n[0] -= '0';
    n[1] -= '0';
    if (n[0] > 9) {
      if (n[0] > 'F' - '0')
        n[0] -= 'a' - '9' - 1;
      else
        n[0] -= 'A' - '9' - 1;
    }
    if (n[1] > 9) {
      if (n[1] > 'F' - '0')
        n[1] -= 'a' - '9' - 1;
      else
        n[1] -= 'A' - '9' - 1;
    }

    /* store as octets */
    x[i]  = n[0] << 4;
    x[i] += n[1];
  } /* for (each octet) */

  return l/2;
}

/*
 * Convert an ASCII keystring to a keyblock.
 * Returns keylen on success, -1 otherwise.
 */
static int
keystring2keyblock(const char *keystring, unsigned char keyblock[])
{
  size_t l = strlen(keystring);

  /* 160-bit key with optional line ending */
  if ((l & ~1) != 40)
    return 1;

  return a2x(keystring, keyblock);
}

static char *
x2a(char s[17], unsigned char challenge[8])
{
  int i;

  for (i = 0; i < 8; i++)
    (void) sprintf(&s[i*2], "%02x", challenge[i]);
  return s;
}

static void
hotp(unsigned char challenge[], unsigned char keyblock[],
     unsigned char response[])
{
  uint32_t dbc;		/* "dynamic binary code" from HOTP draft */
  unsigned char md[20];
  unsigned md_len;

  /* 1. hmac */
  (void) HMAC(EVP_sha1(), keyblock, 20, challenge, 8, md, &md_len) ;

  /* 2. the truncate step is unnecessarily complex */
  {
    int offset;

    offset = md[19] & 0x0F;
    /* we can't just cast md[offset] because of alignment and endianness */
    dbc = (md[offset] & 0x7F) << 24 |
           md[offset + 1]     << 16 |
           md[offset + 2]     << 8  |
           md[offset + 3];
  }

  /* 3. int conversion and modulus (as string) */
  (void) sprintf((char *) response, "%06lu", dbc % 1000000L);
}

static void
c2c(uint64_t counter, unsigned char challenge[8])
{
  challenge[0] = counter >> 56;
  challenge[1] = counter >> 48;
  challenge[2] = counter >> 40;
  challenge[3] = counter >> 32;
  challenge[4] = counter >> 24;
  challenge[5] = counter >> 16;
  challenge[6] = counter >> 8;
  challenge[7] = counter;
}

int
main(int argc, char *argv[])
{
  char *username = "username";
  char *keystring = "0000000000000000000000000000000000000000";
  char *pass1 = NULL;
  char *pass2 = NULL;
  unsigned char challenge[8];
  unsigned char response[7];
  unsigned char keyblock[20];
  uint64_t counter;
  uint64_t initial = 0;
  uint64_t final = 0;

  int debug = 0;

  {
    int c;

    while ((c = getopt(argc, argv, ":d1:2:u:k:c:i:f:")) != -1) {
      switch (c) {
        case 'd': debug = 1; break;
        case '1': pass1 = optarg; break;
        case '2': pass2 = optarg; break;
        case 'u': username = optarg; break;
        case 'k': keystring = optarg; break;
        case 'c': initial = strtoull(optarg, NULL, 16); break;
        case 'i': initial = strtoull(optarg, NULL, 10); break;
        case 'f':
          if (optarg[0] == '+')	/* NOTE: must be given AFTER -i */
            final = initial + strtoull(&optarg[1], NULL, 10);
          else
            final = strtoull(optarg, NULL, 10);
          break;
        default:
          (void) fprintf(stderr, "%s: unknown option\n", argv[0]);
          exit(1);
      }
    }
  }
  if (!pass1 || !pass2)
    usage(argv[0]);
  if (strlen(pass1) != 6) {
    (void) fprintf(stderr, "%s: passcode 1 wrong length\n", argv[0]);
    exit(1);
  }
  if (strlen(pass2) != 6) {
    (void) fprintf(stderr, "%s: passcode 2 wrong length\n", argv[0]);
    exit(1);
  }

  if (!final)
    final = initial + 65536;

  (void) keystring2keyblock(keystring, keyblock);
  (void) memset(challenge, 0, 8);

  for (counter = initial; counter < final; ++counter) {
    c2c(counter, challenge);
    hotp(challenge, keyblock, response);
   
    if (debug == 1)
       (void) printf("%llu: %s\n", (long long unsigned int) counter, response);

    if (!strcmp((char *) response, pass1)) {
      if (debug)
        (void) printf("matched %s at counter=%llu\n", pass1, (long long unsigned int) counter);

      /* matched first pass, look for 2nd */
      c2c(++counter, challenge);
      hotp(challenge, keyblock, response);
      if (!strcmp((char *) response, pass2)) {
        char s[17];

        /* token is now at the subsequent counter value */
        c2c(++counter, challenge);
        /* print the new state */
        (void) printf("5:%s:%s:::0:0:0:\n", username, x2a(s, challenge));
        exit(0);

      } else if (debug) {
        (void) fprintf(stderr,
                       "mismatch for counter=%llu, wanted: %s got: %s\n",
                       (long long unsigned int) counter, pass2, response);
        (void) fprintf(stderr, "continuing search\n");
      } /* else (pass2 does not match) */
    } /* if (pass1 matches) */
  } /* for (each event counter) */

  (void) fprintf(stderr, "%s: counter value not found\n", argv[0]);
  exit(1);
}
